### ISSUE TITLE
Strange behavior with functools.cached_property

### ISSUE THREAD

**Issue by tjf801** [microsoft/pyright#4957]
**Describe the bug**
Using cached properties doesn't work in certain cases when using a normal property works fine.

**Screenshots or Code**
```python
# pyright: strict

from functools import cached_property

class Base:
    @cached_property
    def foo(self) -> 'Base':
        ...

class Derived(Base):
    @cached_property
    def foo(self) -> 'Derived':
        ...

```
It should also be noted that this example works fine when the type checking mode is not strict, or when the `cached_property` decorators are replaced with `property`. 

**VS Code extension or command-line**
I am using the current pylance VS code extension.

**Additional context**
It should be noted that in my actual use case here that `typing.Self` can't be used as the return type, because I have a derived class that returns an instance of a different derived class.

**Comment by erictraut** [comment]
This is related to https://github.com/microsoft/pyright/issues/4343.

This behavior is an unfortunate side effect of how `cached_property` is defined in typeshed's `functools.pyi` stub. The type parameter for the `cached_property` class is invariant. It should probably be covariant, which would fix this issue. Feel free to report the problem to the maintainers of typeshed.
